### PR TITLE
RFC: Session/Error Feeds ClickHouse Migration

### PR DESCRIPTION
**Pull request by Vadman97**
## Summary

Adds an RFC for migrating opensearch usage to clickhouse.

### CODE REVIEW

File: internal-docs/rfcs/002-feed-clickhouse-migration.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+## Database modeling
+
+As a starting point, the database schema can be similar to the current postgres schema for sessions and errors. 
+However, we'll need to benchmark the performance of different approaches for querying sessions and errors in ClickHouse. 
+Ie.
+1. having 3 separate tables matching what’s stored in Postgres and join them at query time
+2. denormalize fields and save a copy of their values with each session/error (as we do with OpenSearch)
+3. treat queryable session/error columns as fields too and denormalize with a format like (field_type, field_name, field_value, id), then look up the full session/error objects in postgres after getting ids from clickhouse
+
+Migrating data should be straightforward with a ClickHouse query as per [this blog](https://clickhouse.com/blog/migrating-data-between-clickhouse-postgres).
```
Comment (et) [inline internal-docs/rfcs/002-feed-clickhouse-migration.md:33]: There's actually a couple [strategies](https://clickhouse.com/blog/migrating-data-between-clickhouse-postgres) in here. When I looked at this, it seems to figure out how to populate a Clickhouse db but there wasn't good information on how to keep this information in sync (i.e. forward writes).

With OpenSearch, we keep data in sync through application code. Would we want the same approach here?

File: internal-docs/rfcs/002-feed-clickhouse-migration.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+* We'll be writing new types of ClickHouse aggregation queries which we are not as familiar with.
+* However, we'll be able to deprecate OpenSearch and save time currently spent when looking at performance issues.
+
+## Open Questions
+
+Since ClickHouse [does not support high-volume updates](https://clickhouse.com/docs/en/guides/developer/mutations) to rows, we would not want to use ClickHouse as the source of truth for session updated / processed states. However, we could re-architect our session processing logic to store that frequently-modified data in Redis while storing final records in ClickHouse.
+
+## Rollout + Adoption
+
+1. Migrate sessions and error groups from PostgreSQL to ClickHouse by using the [cross-db insert workflow](https://clickhouse.com/blog/migrating-data-between-clickhouse-postgres).
```
Comment (et) [inline internal-docs/rfcs/002-feed-clickhouse-migration.md:59]: What is `cross-db insert workflow`?

**Review by et — COMMENTED** [review]
Thanks for driving this forward.

**Review by deltaepsilon — APPROVED** [review]

File: internal-docs/prds/002-feed-clickhouse-migration.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  - supports exact and like matching for attributes + and/or logic between attributes
+  - shows a histogram of sessions over time for the current query, grouped by an attribute (default - has/no-has errors)
+  - new: for each session, shows a distribution-preview of events over time (grouping by errors, track events, etc.)
+- Command bar
+  - Searching for sessions and errors based on predefined attributes: email, identifier, url, os, browser
+  - new: using the input to predict the type of query that makes more sense (session vs. error vs. log)
+
+## User Feedback
+
+This feature is motivated by poor performance of OpenSearch queries (cold-start query start can take >5 seconds)
```
[the remark this answers is not in the document]
Comment (Vadman97) [reply internal-docs/prds/002-feed-clickhouse-migration.md:35]: "cold start" means visiting app.highlight.io/1/sessions or /errors during off hours when opensearch queries haven't been performed in a while. we've found opensearch to be faster on a follow-up query compared to a query after a period of inactivity. cc @mayberryzane

File: internal-docs/prds/002-feed-clickhouse-migration.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+and the growing costs of storing all session and error metadata in OpenSearch.
```
[the remark this answers is not in the document]
Comment (Vadman97) [reply internal-docs/prds/002-feed-clickhouse-migration.md:36]: $7k for OpenSearch compared to ~$1.5k estimated for ClickHouse

File: internal-docs/rfcs/002-feed-clickhouse-migration.md (hunk 2 above)
Comment (Vadman97) [reply internal-docs/rfcs/002-feed-clickhouse-migration.md:59]: sorry, i thought the linked blog post made this clear enough.
see the https://clickhouse.com/blog/migrating-data-between-clickhouse-postgres#postgres-to-clickhouse section
```INSERT INTO uk_price_paid SELECT * EXCEPT id FROM postgresql('...') ...```

File: internal-docs/rfcs/002-feed-clickhouse-migration.md (hunk 1 above)
Comment (Vadman97) [reply internal-docs/rfcs/002-feed-clickhouse-migration.md:33]: correct, we'd want to do dual writes via application code until we achieve feature parity, at which point we would remove opensearch writes. i'll clarify that here.

File: internal-docs/rfcs/002-feed-clickhouse-migration.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+Migrating data should be straightforward with a ClickHouse query as per [this blog](https://clickhouse.com/blog/migrating-data-between-clickhouse-postgres).
+We can start by copying the current PostgreSQL schema and determining what the ClickHouse SQL queries replacing the OpenSearch queries will look like.
+We can then create a copy of the error groups table in ClickHouse which denormalizes the fields into a map column type, and benchmark the updated queries.
+Lastly, we can create another copy to see if the third schema is a performant option. The benefit of option #3 is that postgres can continue to be
+the source of truth for error group data, but we should consider if ClickHouse can be the long-term source of truth for write-once metadata.
+
+## Success Metrics
+
+* The features listed in the [PRD](../prds/002-feed-clickhouse-migration.md) are implemented.
+* The performance of session, errors, and command bar searches is better than currently with OpenSearch. A goal here is for all worst-case queries to take less than 1 second.
```
[the remark this answers is not in the document]
Comment (Vadman97) [reply internal-docs/rfcs/002-feed-clickhouse-migration.md:42]: yup - https://app.highlight.io/1/dashboards/2

image

**Pull request merged**